staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -93,6 +93,10 @@ func (t *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	resultCh := make(chan interface{})
 	var tw timeoutWriter
 	tw, w = newTimeoutWriter(w)
+
+	//Make a copy of request and work on it in new goroutine
+	//to avoid race condition when accessing/modifying request headers
```

Comment:
Are request headers the only problematic case?

If so, maybe we can shallow copy request and deep clone headers only to save few bytes?

Reply:
I think for now we believe only headers are problematic, but I can imagine (e.g. in future) some handler modifying any other field in http.Request.

I think it's reasonable to clone entire object, but I would change the comment to say something like "to avoid races when accessing/modifying http.Request (e.g. request headers)".

Reply:
Yes, I would rather make deep copy of all request metadata. This way we ensure that further changes to filter that precede timeout handler  probably won't cause race condition. I think there is very low probability that any filter will touch body of the request, which is not deep copied here.